associated-token-account/program/src/processor.rs

```diff
@@ -56,6 +70,21 @@ pub fn process_create_associated_token_account(
     let spl_token_program_info = next_account_info(account_info_iter)?;
     let spl_token_program_id = spl_token_program_info.key;
 
+    if create_mode == CreateMode::IfNonExistent
+        && associated_token_account_info.owner == spl_token_program_id
```

Comment:
I think we should explicitly return `ProgramError::IllegalOwner` in the contra case for the sake of hardening against future errors

Reply:
Sorry, _errors_ is ambiguous here. I mean _programming_ errors :wink:

Reply:
Which contra case do you mean exactly?  If `associated_token_account_info.owner != spl_token_program_id`, that's still valid, because it'll be `system_program::id()` if the account hasn't been created. Do you mean for it to go:
```
if create_mode == ... {
    if ata.owner == spl_token_program_id {
        ...
    } else if ata.owner != system_program::id() {
        return Err(ProgramError::IllegalOwner);
    }
}
```

Reply:
Yep! That's the logic I'd like to make prominent